### PR TITLE
Update EulerAngle after rigid body rotation

### LINKED ISSUE #7959: Update grain orientation after rigid body rotation
### Description of the enhancement or error report

Grain orientation should be updated at each time step when rigid body rotation is considered. 
### Rationale for the enhancement or information for reproducing the error

Additional EulerAngleProvider class to update angles depending on torque responsible for grain rotation.
### Identified impact

(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

No impact on existing codes.

### PR DESCRIPTION
**Pull request by SudiptaBiswas**
- Adds additional `EulerAngleProvider` class for updating the grain orientations after rigid body rotation
- closes #7959

### CODE REVIEW

**Comment by dschwen** [comment]
Mammoth and Rattlesnake have (unrelated) timeouts.

**Comment by permcody** [comment]
@dschwen - poke

File: modules/phase_field/include/vectorpostprocessors/EulerAngleUpdaterCheck.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+class EulerAngleProvider;
+class GrainTrackerInterface;
+class GrainForceAndTorqueInterface;
+class RotationTensor;
+
+template<>
+InputParameters validParams<EulerAngleUpdaterCheck>();
+
+/**
+ * Update Euler angles of each grains after rigid body rotation
```
Comment (dschwen) [inline modules/phase_field/include/vectorpostprocessors/EulerAngleUpdaterCheck.h]: Can you document what _this_ class does.

File: modules/phase_field/src/userobjects/EulerAngleUpdater.C [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  }
+
+  unsigned int angle_size = _angles.size();
+  for (unsigned int i = angle_size; i < grain_num; ++i)
+    _angles.push_back(_euler.getEulerAngles(i));
+
+  for (unsigned int i = 0; i < grain_num; ++i)
+  {
+    if (!_first_time)
+      if (!_fe_problem.converged())
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/EulerAngleUpdater.C]: do `if (!_first_time && !_fe_problem.converged())`

File: modules/phase_field/src/userobjects/EulerAngleUpdater.C [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      _angles_old[i] = _angles[i];
+    else
+      _angles_old.push_back(_angles[i]);
+
+    RotationTensor R0(_angles_old[i]);
+    RealVectorValue torque_rotated = R0 * torque;
+    RealVectorValue omega = _mr / _grain_volumes[i] * torque_rotated;
+
+    RealVectorValue angle_change;
+    angle_change(0) = omega(2) * _dt;
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/EulerAngleUpdater.C]: Please add a comment here, what are you trying to do here. Are you just building a rotation matrix for the torque effect?

File: modules/phase_field/src/userobjects/EulerAngleUpdater.C [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    angle_change(1) = (omega(0) * std::cos(angle_change(0)) + omega(1) * std::sin(angle_change(0))) * _dt;
+    angle_change(2) = (omega(0) * std::sin(angle_change(0)) * std::sin(angle_change(1))
+                       - omega(1) * std::cos(angle_change(0)) * std::sin(angle_change(1))
+                       + omega(2) * std::cos(angle_change(1))) * _dt;
+    angle_change *= (180.0 / libMesh::pi);
+
+    RotationTensor R1(angle_change);
+    RealTensorValue R = R1 * R0;
+
+    if (R(2,2) != 1.0 && R(2,2) != -1.0)
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/EulerAngleUpdater.C]: Add comment. You are back tracking the Euler angles out of a rotation matrix here.

File: modules/phase_field/src/vectorpostprocessors/EulerAngleUpdaterCheck.C [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    RotationTensor R(_angles[i]);
+    RealVectorValue a_rot = R * a;
+
+    RotationTensor R0(_angles_old[i]);
+    RealVectorValue torque_rot = R0 * torque;
+    RealVectorValue a_rot0 = R0 * a;
+
+    RealVectorValue torque_rot1;
+    RealVectorValue angle_rot;
+    torque_rot1(0) = torque_rot(2);
```
Comment (dschwen) [inline modules/phase_field/src/vectorpostprocessors/EulerAngleUpdaterCheck.C]: comments, comments, comments, please

**Review by dschwen — DISMISSED** [review]

**Comment by permcody** [comment]
Looks like this PR is ready for another review.

File: modules/phase_field/include/userobjects/EulerAngleUpdater.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * Update Euler angles of each grains after rigid body rotation
+ * This class estimates the rotation of principal axes of the grains due to applied torques
+ * and calculates the final grain orientation.
+ * Step1: Calculate RotationTensor based on euler angles from previous time step (R0)
+ * Step2: Obtain the torque acting on the grains at current time Step
+ * Step3: Calculate the angular velocities around global axes
+ * Step4: Calculae change in euler angles due to torque and corresponding rotation matrix(R1)
+ * Step5: Calculae final rotation matrix, R = R1 * R0, determines the final position of any rotatted vector
```
Comment (dschwen) [inline modules/phase_field/include/userobjects/EulerAngleUpdater.h]: 2x Calculae

Comment (dschwen) [inline modules/phase_field/include/userobjects/EulerAngleUpdater.h]: and rotatted

File: modules/phase_field/include/vectorpostprocessors/EulerAngleUpdaterCheck.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+class EulerAngleProvider;
+class GrainTrackerInterface;
+class GrainForceAndTorqueInterface;
+class RotationTensor;
+
+template<>
+InputParameters validParams<EulerAngleUpdaterCheck>();
+
+/**
+ * This is an unit test to check the correctness of the updated euler angles
```
Comment (dschwen) [inline modules/phase_field/include/vectorpostprocessors/EulerAngleUpdaterCheck.h]: a unit

File: modules/phase_field/src/userobjects/EulerAngleUpdater.C [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+    RealGradient torque = _grain_torque.getTorqueValues()[i];
+
+    if (i <= angle_size) // if new grains are created
+      _angles_old[i] = _angles[i];
+    else
+      _angles_old.push_back(_angles[i]);
+
+    RotationTensor R0(_angles_old[i]); // RotationTensor as per old euler angles
+    RealVectorValue torque_rotated = R0 * torque; // Applied torque is rotated to alligh with old grain axes
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/EulerAngleUpdater.C]: alligh

**Review by dschwen — COMMENTED** [review]
Some typos in comments. Looks good otherwise (thanks for adding all the comments!)

**Comment by permcody** [comment]
@SudiptaBiswas - Could I get you to rebase this commit once more and repush? We just merged in a really big PR. I would prefer to retest some of these PRs.

**Comment by SudiptaBiswas** [comment]
This is updated now.

**Comment by permcody** [comment]
Noted

**Review by permcody — APPROVED** [review]
Looks good, thanks for rebasing.

**Pull request merged**